### PR TITLE
Add timing metrics to puppet-run.

### PR DESCRIPTION
**Pull request by alienth**
@kylebrandt @captncraig

### CODE REVIEW

**Comment by alienth** [comment]
Output from test:

```
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.519263,"tags":{"host":"centos1","time":"Augeas"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":5.47541999816895,"tags":{"host":"centos1","time":"ConfigRetrieval"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.001653,"tags":{"host":"centos1","time":"Cron"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":2.823981,"tags":{"host":"centos1","time":"Exec"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":2.449183,"tags":{"host":"centos1","time":"File"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.000242,"tags":{"host":"centos1","time":"Filebucket"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.019257,"tags":{"host":"centos1","time":"Group"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.013445,"tags":{"host":"centos1","time":"IniSetting"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":1.71764,"tags":{"host":"centos1","time":"Package"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.001041,"tags":{"host":"centos1","time":"Schedule"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":1.616973,"tags":{"host":"centos1","time":"Service"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.009557,"tags":{"host":"centos1","time":"SshAuthorizedKey"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.028535,"tags":{"host":"centos1","time":"User"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.012581,"tags":{"host":"centos1","time":"Yumrepo"}}
2015/02/25 13:41:58 info: queue.go:66: {"metric":"puppet.run_duration_total","timestamp":1424902362,"value":14.6904529981689,"tags":{"host":"centos1"}}
```

**Comment by alienth** [comment]
Note: collector is not configured to fetch 'file_line' and 'vcsrepo' timings from summary report.

If we do want those timings in there as well, let me know and I'll add them to the struct.

**Comment by alienth** [comment]
Abandoned my original approach, as it required hardcoded module names (which are likely to change).

New approach will send timings for whatever module exists.

**Comment by alienth** [comment]
Test output:

```

2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.disabled","timestamp":1424915229,"value":0,"tags":{"host":"centos1"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.disabled_no_reason","timestamp":1424915229,"value":0,"tags":{"host":"centos1"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run.resources","timestamp":1424902362,"value":2,"tags":{"host":"centos1","resource":"changed"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run.resources","timestamp":1424902362,"value":0,"tags":{"host":"centos1","resource":"failed"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run.resources","timestamp":1424902362,"value":0,"tags":{"host":"centos1","resource":"failed_to_restart"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run.resources","timestamp":1424902362,"value":2,"tags":{"host":"centos1","resource":"out_of_sync"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run.resources","timestamp":1424902362,"value":0,"tags":{"host":"centos1","resource":"restarted"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run.resources","timestamp":1424902362,"value":0,"tags":{"host":"centos1","resource":"scheduled"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run.resources","timestamp":1424902362,"value":2,"tags":{"host":"centos1","resource":"skipped"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run.resources_total","timestamp":1424902362,"value":329,"tags":{"host":"centos1"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run.changes","timestamp":1424902362,"value":2,"tags":{"host":"centos1"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.012581,"tags":{"host":"centos1","time":"yumrepo"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.000358,"tags":{"host":"centos1","time":"file_line"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.028535,"tags":{"host":"centos1","time":"user"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.013445,"tags":{"host":"centos1","time":"ini_setting"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.519263,"tags":{"host":"centos1","time":"augeas"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.009557,"tags":{"host":"centos1","time":"ssh_authorized_key"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":1.71764,"tags":{"host":"centos1","time":"package"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.000398,"tags":{"host":"centos1","time":"vcsrepo"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":2.449183,"tags":{"host":"centos1","time":"file"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.000926,"tags":{"host":"centos1","time":"sshkey"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.001041,"tags":{"host":"centos1","time":"schedule"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":1.616973,"tags":{"host":"centos1","time":"service"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.019257,"tags":{"host":"centos1","time":"group"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":5.47541999816895,"tags":{"host":"centos1","time":"config_retrieval"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":2.823981,"tags":{"host":"centos1","time":"exec"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.001653,"tags":{"host":"centos1","time":"cron"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration_total","timestamp":1424902362,"value":14.6904529981689,"tags":{"host":"centos1"}}
2015/02/25 16:47:09 info: queue.go:66: {"metric":"puppet.run_duration","timestamp":1424902362,"value":0.000242,"tags":{"host":"centos1","time":"filebucket"}}
```

File: cmd/scollector/collectors/puppet_linux.go [hunk 1]
```diff
@@ -95,15 +78,28 @@ func puppet_linux() (opentsdb.MultiDataPoint, error) {
 	if err = yaml.Unmarshal(s, &m); err != nil {
 		return nil, err
 	}
+
```
Comment (kylebrandt) [inline cmd/scollector/collectors/puppet_linux.go]: Remove Empty line -- We don't have them within funcs

File: cmd/scollector/collectors/puppet_linux.go [hunk 2]
```diff
@@ -95,15 +78,28 @@ func puppet_linux() (opentsdb.MultiDataPoint, error) {
...
+	AddTS(&md, "puppet.run.resources", last_run, m.Resources.Failed, opentsdb.TagSet{"resource": "failed"}, metadata.Gauge, metadata.Count, "")
+	AddTS(&md, "puppet.run.resources", last_run, m.Resources.FailedToRestart, opentsdb.TagSet{"resource": "failed_to_restart"}, metadata.Gauge, metadata.Count, "")
+	AddTS(&md, "puppet.run.resources", last_run, m.Resources.OutOfSync, opentsdb.TagSet{"resource": "out_of_sync"}, metadata.Gauge, metadata.Count, "")
+	AddTS(&md, "puppet.run.resources", last_run, m.Resources.Restarted, opentsdb.TagSet{"resource": "restarted"}, metadata.Gauge, metadata.Count, "")
+	AddTS(&md, "puppet.run.resources", last_run, m.Resources.Scheduled, opentsdb.TagSet{"resource": "scheduled"}, metadata.Gauge, metadata.Count, "")
+	AddTS(&md, "puppet.run.resources", last_run, m.Resources.Changed, opentsdb.TagSet{"resource": "skipped"}, metadata.Gauge, metadata.Count, "")
+	AddTS(&md, "puppet.run.resources_total", last_run, m.Resources.Total, nil, metadata.Gauge, metadata.Count, "")
+	AddTS(&md, "puppet.run.changes", last_run, m.Changes.Total, nil, metadata.Gauge, metadata.Count, "")
+
```
Comment (kylebrandt) [inline cmd/scollector/collectors/puppet_linux.go]: Remove Empty line here as well

File: cmd/scollector/collectors/puppet_linux.go [hunk 3]
```diff
@@ -95,15 +78,28 @@ func puppet_linux() (opentsdb.MultiDataPoint, error) {
...
+	for k, v := range m.Time {
+		metric, ok := v.(float64)
+		if ok {
+			if k == "total" {
+			        AddTS(&md, "puppet.run_duration_total", last_run, metric, nil, metadata.Gauge, metadata.Second, "")
```
Comment (kylebrandt) [inline cmd/scollector/collectors/puppet_linux.go]: Best practice here is too add description strings (Metadata was added midlife). Descriptions are full sentences than end with a period, and are usually stored in constants (i.e. see exim_linux.go). These descriptions show up on the graph page, and can be included in alert notifications. Also one of the issues I put in as a feature request is to make the desc information searchable when looking for metrics.

**Comment by alienth** [comment]
@kylebrandt  requested changes made. Also caught an existing bug.

**Comment by alienth** [comment]
Replaced the interface{} usage with strings, and now using strconv.

**Pull request merged**